### PR TITLE
Adding Password History UI improvements

### PR DESCRIPTION
**Pull request by GayanM**

### CODE REVIEW

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 1]
```diff
@@ -131,11 +132,27 @@ public void updatePassword(String username, char[] oldPassword, char[] newPasswo
...
+    /**
+     * This will be removed with new config model
+     * @return
+     */
+    public int getHistoryCount () {
+        return new PasswordHistoryBean().getMinCountToAllowRepitition();
+    }
+
+    /**
+     * This will be removed with new config model
```
Comment (Pushpalanka) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: Better to add a ToDo item for this.

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js [hunk 2]
```diff
@@ -24,6 +24,25 @@ function updatePassword(username, oldPassword, newPassword, domain) {
...
+        if (cause != null) {
+            //the exceptions thrown by the actual osgi service method is wrapped inside a InvocationTargetException.
+            if (cause instanceof java.lang.reflect.InvocationTargetException) {
+                message = cause.getTargetException().message;
+        		errorCode = cause.getTargetException().getErrorCode();
+
+        		if (errorCode === "1000") {
+                    var count = callOSGiService("org.wso2.is.portal.user.client.api.IdentityStoreClientService",
+                            "getHistoryCount", []);
+                    message = "Password You entered has already been used in last " + count + "Attempts";
```
Comment (Pushpalanka) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js]: I assume this error shown to the end user. In that case this need to be properly internationalized. Better to have message as "Entered password has already been used in last x days."  Attention capital, simple letter usage too.

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js [hunk 3]
```diff
@@ -24,6 +24,25 @@ function updatePassword(username, oldPassword, newPassword, domain) {
...
+                } else if(errorCode === "1001") {
+                    var days = callOSGiService("org.wso2.is.portal.user.client.api.IdentityStoreClientService",
+                    "getNumOfDays", []);
+                    message = "Password You entered has already been used in last " + days;
```
Comment (Pushpalanka) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js]: Same above comment applies here.

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js [hunk 4]
```diff
@@ -24,7 +24,26 @@ function updatePassword(username, oldPassword, newPassword, domain) {
         return {success: true, message: "You have successfully updated the password"};
     } catch (e) {
         var message = e.message;
-        return {success: false, message: message};
+        var errorCode = -1;
+        var cause = e.getCause();
+        if (cause != null) {
+            //the exceptions thrown by the actual osgi service method is wrapped inside a InvocationTargetException.
+            if (cause instanceof java.lang.reflect.InvocationTargetException) {
+                message = cause.getTargetException().message;
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js]: Fix formatting

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js [hunk 5]
```diff
@@ -24,7 +24,26 @@ function updatePassword(username, oldPassword, newPassword, domain) {
         return {success: true, message: "You have successfully updated the password"};
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js]: message should support internationalization

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js [hunk 6]
```diff
@@ -21,10 +21,29 @@ function updatePassword(username, oldPassword, newPassword, domain) {
         callOSGiService("org.wso2.is.portal.user.client.api.IdentityStoreClientService",
             "updatePassword", [username, oldPasswordChar, newPasswordChar, domain]);
 
-        return {success: true, message: "You have successfully updated the password"};
+        return {success: true, message: 'password.history.validation.success'};
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js]: @Pushpalanka i18 string should be changed to "password.update.success"

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js [hunk 7]
```diff
@@ -21,10 +21,29 @@ function updatePassword(username, oldPassword, newPassword, domain) {
...
     } catch (e) {
         var message = e.message;
-        return {success: false, message: message};
+        var errorCode = -1;
+        var cause = e.getCause();
+        if (cause !== null) {
+            //the exceptions thrown by the actual osgi service method is wrapped inside a InvocationTargetException.
+            if (cause instanceof java.lang.reflect.InvocationTargetException) {
+                message = cause.getTargetException().message;
+                errorCode = cause.getTargetException().getErrorCode();
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/fragments/reset-password/reset-password.js]: check instance of UserPortalUIException

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/lang/en_US.properties [hunk 8]
```diff
@@ -169,3 +169,6 @@ code=Code
 back.sign.in=Back to sign in
 recover.username=Recover Username
 fill.below.to.recover=Please fill in the below form to recover your username
+password.history.validation.error.1000=Password you entered has already been used in last N attempts"
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/lang/en_US.properties]: {}

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java [hunk 9]
```diff
@@ -131,11 +132,27 @@ public void updatePassword(String username, char[] oldPassword, char[] newPasswo
         } catch (IdentityStoreException e) {
             String error = "Failed to update user password.";
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java]: add context information to error message

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/PasswordHistoryBean.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Bean that encapsulates the configuration info.
+ */
+public class PasswordHistoryBean implements Serializable {
+
+    private static final long serialVersionUID = -2913500114444797062L;
+    private boolean isEnabled = true;
+    private int minCountToAllowRepitition = 5;
+    private int minAgeToAllowRepitition = 7;
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/PasswordHistoryBean.java]: change to minAgeToAllowRepitition -> minDaysToAllowRepitition

**Pull request merged**